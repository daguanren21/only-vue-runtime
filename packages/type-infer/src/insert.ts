import type { t } from './babel'
import type { RuntimeProps } from './types'
import fs from 'fs-extra'
import MagicString from 'magic-string'
import { parse, traverse } from './babel'

const map: Record<string, string> = {
  TSArrayType: 'Array',
  TSTypeLiteral: 'Object',
  TSNumberKeyword: 'Number',
  TSStringKeyword: 'String',
  TSBooleanKeyword: 'Boolean',
}

function mapTsType2VueType(type: string) {
  return map[type] || type
}

export function insert(id: string, props: RuntimeProps) {
  const code = fs.readFileSync(id, 'utf8')
  const ms = new MagicString(code)
  const ast = parse(code, {
    sourceType: 'module',
  })

  traverse(ast, {
    CallExpression(p) {
      if (
        p.get('callee').isIdentifier({
          name: '_defineComponent',
        })
      ) {
        const options = p.get('arguments')[0]
        if (options.isObjectExpression()) {
          const propertiesPath = options.get('properties')
          const propsKey = propertiesPath.find(p => p.get('key').isIdentifier({
            name: 'props',
          }))
          if (!propsKey) {
            const nameKey = propertiesPath.find(p => p.get('key').isIdentifier({
              name: '__name',
            }))
            if (nameKey && typeof nameKey.node.start === 'number' && typeof nameKey.node.end === 'number') {
              const padLeft = ' '.repeat(nameKey.node.loc?.start.column ?? 0)
              ms.appendRight(nameKey.node.end + 1, [
                `props: {`,
                ...Object.entries(props).map(([key, { type, required }]) => {
                  return `  ${key}: { type: ${mapTsType2VueType(type)}, required: ${required} },`
                }),
                '},',
              ].map((x, idx) => {
                return `${idx === 0 ? '\n' : ''}${padLeft}${x}`
              }).join('\n'))
            }
          }
        }
      }
    },
  })

  return ms.toString()
}
