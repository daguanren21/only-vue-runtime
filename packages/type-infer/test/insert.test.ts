import { insert } from '@/insert'
import fs from 'fs-extra'
import path from 'pathe'

function handle(jsonFilename: string) {
  const code = insert(
    path.resolve(__dirname, `./fixtures/vue/index.js`),
    fs.readJsonSync(path.resolve(__dirname, `./output/${jsonFilename}.json`)),
  )
  // fs.outputFileSync(path.resolve(__dirname, `./output/${jsonFilename}.js`), code)
  return code
}

describe('insert', () => {
  it('simple-interface-ref', () => {
    expect(handle('simple-interface-ref')).toMatchSnapshot()
  })

  it('simple-interface', () => {
    expect(handle('simple-interface')).toMatchSnapshot()
  })

  it('simple-type-ref-interseaction-case-0', () => {
    expect(handle('simple-type-ref-interseaction-case-0')).toMatchSnapshot()
  })

  it('simple-type-ref', () => {
    expect(handle('simple-type-ref')).toMatchSnapshot()
  })
})
