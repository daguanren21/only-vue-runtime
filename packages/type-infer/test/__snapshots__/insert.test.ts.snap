// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`insert > simple-interface 1`] = `
"import { defineComponent as _defineComponent } from 'vue'
import { ref } from 'vue'


const __sfc__ = /*@__PURE__*/_defineComponent({
  __name: 'App',
  props: {
    a: { type: Number, required: true },
    b: { type: String, required: false },
    c: { type: Boolean, required: true },
    d: { type: Array, required: true },
    e: { type: Object, required: true },
  },
  setup(__props, { expose: __expose }) {
  __expose();

const msg = ref('Hello World!')




const __returned__ = { msg }
Object.defineProperty(__returned__, '__isScriptSetup', { enumerable: false, value: true })
return __returned__
}

});
import { toDisplayString as _toDisplayString, createElementVNode as _createElementVNode, vModelText as _vModelText, withDirectives as _withDirectives, Fragment as _Fragment, openBlock as _openBlock, createElementBlock as _createElementBlock } from "vue"
function render(_ctx, _cache, $props, $setup, $data, $options) {
  return (_openBlock(), _createElementBlock(_Fragment, null, [
    _createElementVNode("h1", null, _toDisplayString($setup.msg), 1 /* TEXT */),
    _withDirectives(_createElementVNode("input", {
      "onUpdate:modelValue": _cache[0] || (_cache[0] = $event => (($setup.msg) = $event))
    }, null, 512 /* NEED_PATCH */), [
      [_vModelText, $setup.msg]
    ])
  ], 64 /* STABLE_FRAGMENT */))
}
__sfc__.render = render
__sfc__.__file = "src/App.vue"
export default __sfc__"
`;

exports[`insert > simple-interface-ref 1`] = `
"import { defineComponent as _defineComponent } from 'vue'
import { ref } from 'vue'


const __sfc__ = /*@__PURE__*/_defineComponent({
  __name: 'App',
  props: {
    a: { type: Number, required: true },
    b: { type: String, required: false },
    c: { type: Boolean, required: true },
    d: { type: Array, required: true },
    e: { type: Object, required: true },
  },
  setup(__props, { expose: __expose }) {
  __expose();

const msg = ref('Hello World!')




const __returned__ = { msg }
Object.defineProperty(__returned__, '__isScriptSetup', { enumerable: false, value: true })
return __returned__
}

});
import { toDisplayString as _toDisplayString, createElementVNode as _createElementVNode, vModelText as _vModelText, withDirectives as _withDirectives, Fragment as _Fragment, openBlock as _openBlock, createElementBlock as _createElementBlock } from "vue"
function render(_ctx, _cache, $props, $setup, $data, $options) {
  return (_openBlock(), _createElementBlock(_Fragment, null, [
    _createElementVNode("h1", null, _toDisplayString($setup.msg), 1 /* TEXT */),
    _withDirectives(_createElementVNode("input", {
      "onUpdate:modelValue": _cache[0] || (_cache[0] = $event => (($setup.msg) = $event))
    }, null, 512 /* NEED_PATCH */), [
      [_vModelText, $setup.msg]
    ])
  ], 64 /* STABLE_FRAGMENT */))
}
__sfc__.render = render
__sfc__.__file = "src/App.vue"
export default __sfc__"
`;

exports[`insert > simple-type-ref 1`] = `
"import { defineComponent as _defineComponent } from 'vue'
import { ref } from 'vue'


const __sfc__ = /*@__PURE__*/_defineComponent({
  __name: 'App',
  props: {
    a: { type: Number, required: true },
    b: { type: String, required: false },
    c: { type: Boolean, required: true },
    d: { type: Array, required: true },
    e: { type: Object, required: true },
  },
  setup(__props, { expose: __expose }) {
  __expose();

const msg = ref('Hello World!')




const __returned__ = { msg }
Object.defineProperty(__returned__, '__isScriptSetup', { enumerable: false, value: true })
return __returned__
}

});
import { toDisplayString as _toDisplayString, createElementVNode as _createElementVNode, vModelText as _vModelText, withDirectives as _withDirectives, Fragment as _Fragment, openBlock as _openBlock, createElementBlock as _createElementBlock } from "vue"
function render(_ctx, _cache, $props, $setup, $data, $options) {
  return (_openBlock(), _createElementBlock(_Fragment, null, [
    _createElementVNode("h1", null, _toDisplayString($setup.msg), 1 /* TEXT */),
    _withDirectives(_createElementVNode("input", {
      "onUpdate:modelValue": _cache[0] || (_cache[0] = $event => (($setup.msg) = $event))
    }, null, 512 /* NEED_PATCH */), [
      [_vModelText, $setup.msg]
    ])
  ], 64 /* STABLE_FRAGMENT */))
}
__sfc__.render = render
__sfc__.__file = "src/App.vue"
export default __sfc__"
`;

exports[`insert > simple-type-ref-interseaction-case-0 1`] = `
"import { defineComponent as _defineComponent } from 'vue'
import { ref } from 'vue'


const __sfc__ = /*@__PURE__*/_defineComponent({
  __name: 'App',
  props: {
    a: { type: Number, required: true },
    b: { type: String, required: false },
    c: { type: Boolean, required: true },
    d: { type: Array, required: true },
    e: { type: Object, required: true },
    h: { type: Number, required: true },
    i: { type: String, required: false },
    j: { type: Boolean, required: true },
    k: { type: Array, required: true },
    l: { type: Object, required: true },
  },
  setup(__props, { expose: __expose }) {
  __expose();

const msg = ref('Hello World!')




const __returned__ = { msg }
Object.defineProperty(__returned__, '__isScriptSetup', { enumerable: false, value: true })
return __returned__
}

});
import { toDisplayString as _toDisplayString, createElementVNode as _createElementVNode, vModelText as _vModelText, withDirectives as _withDirectives, Fragment as _Fragment, openBlock as _openBlock, createElementBlock as _createElementBlock } from "vue"
function render(_ctx, _cache, $props, $setup, $data, $options) {
  return (_openBlock(), _createElementBlock(_Fragment, null, [
    _createElementVNode("h1", null, _toDisplayString($setup.msg), 1 /* TEXT */),
    _withDirectives(_createElementVNode("input", {
      "onUpdate:modelValue": _cache[0] || (_cache[0] = $event => (($setup.msg) = $event))
    }, null, 512 /* NEED_PATCH */), [
      [_vModelText, $setup.msg]
    ])
  ], 64 /* STABLE_FRAGMENT */))
}
__sfc__.render = render
__sfc__.__file = "src/App.vue"
export default __sfc__"
`;
